iree/compiler/Conversion/LinalgToSPIRV/FlattenMemRefSubspanPass.cpp

```diff
@@ -0,0 +1,329 @@
+    auto divMap = AffineMap::get(0, 2, {sym0.floorDiv(sym1)}, context);
+
+    // Calculate the offset we need to add to the load/store op, in terms of how
+    // many elements.
+    Optional<int64_t> numBytes = getNumBytes(memrefType.getElementType());
+    if (!numBytes) return failure();
+    Value byteValue = rewriter.create<ConstantIndexOp>(op.memref().getLoc(),
+                                                       numBytes.getValue());
+    Value offset = rewriter.create<AffineApplyOp>(
+        op.getLoc(), divMap, ValueRange{subspanOp.byte_offset(), byteValue});
+
+    // Get the new index by adding the old index with the offset.
+    Value newIndex = rewriter.create<AffineApplyOp>(
+        op.getLoc(), addMap, ValueRange{op.indices().front(), offset});
+
+    if (std::is_same<OpType, memref::LoadOp>::value) {
+      rewriter.replaceOpWithNewOp<memref::LoadOp>(
+          op, memrefType.getElementType(), ValueRange{newSubspan, newIndex});
+    } else {
+      rewriter.replaceOpWithNewOp<memref::StoreOp>(
+          op, TypeRange{}, ValueRange{op.getOperand(0), newSubspan, newIndex});
+    }
+
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Pass
+//===----------------------------------------------------------------------===//
+
+struct FlattenMemRefSubspanPass
+    : public PassWrapper<FlattenMemRefSubspanPass, FunctionPass> {
+  FlattenMemRefSubspanPass() {}
+  FlattenMemRefSubspanPass(const FlattenMemRefSubspanPass &pass) {}
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AffineDialect, memref::MemRefDialect>();
+  }
+
+  void runOnFunction() override {
+    // First flatten the dimensions of subspan op and their consumer load/store
+    // ops. This requires setting up conversion targets with type conversion.
+
+    MLIRContext &context = getContext();
+    FlattenMemRefTypeConverter typeConverter;
+    RewritePatternSet flattenPatterns(&context);
+    flattenPatterns.add<FlattenBindingSubspan, LinearizeLoadIndices,
+                        LinearizeStoreIndices>(typeConverter, &context);
+
+    ConversionTarget target(context);
+    target.markUnknownOpDynamicallyLegal([](Operation *) { return true; });
+    target.addDynamicallyLegalOp<memref::LoadOp>([](memref::LoadOp loadOp) {
+      return isRankZeroOrOneMemRef(loadOp.getMemRefType());
+    });
+    target.addDynamicallyLegalOp<memref::StoreOp>([](memref::StoreOp storeOp) {
+      return isRankZeroOrOneMemRef(storeOp.getMemRefType());
+    });
+    target.addDynamicallyLegalOp<IREE::HAL::InterfaceBindingSubspanOp>(
+        [](IREE::HAL::InterfaceBindingSubspanOp subspanOp) {
+          return isRankZeroOrOneMemRef(subspanOp.getType());
+        });
+
+    if (failed(applyFullConversion(getFunction(), target,
+                                   std::move(flattenPatterns)))) {
+      signalPassFailure();
```

Comment:
Do we exit if it failed?

```suggestion
      return signalPassFailure();
```